# Ejercicio 51

Obtener las primeras `n` filas de un objeto `DataFrame`.

## Solución:

In [1]:
import pandas as pd
import numpy as np

In [2]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [3]:
df.iloc[:3]

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí

In [4]:
df.iloc[0:3]

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí

# Ejercicio 52

Obtener las últimas `n` filas de un objeto `DataFrame`.

## Solución:

In [15]:
import pandas as pd
import numpy as np

In [20]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [21]:
df.iloc[-5:]

nombre  puntaje  intentos califico
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
h   Edgar      NaN         1       No
i   Angie      8.0         2       No
j  Irlesa     18.0         1       Sí

# Ejercicio 53

Obtener la representación en **NumPy** de un objeto `DataFrame`.

## Solución:

In [22]:
import pandas as pd

In [23]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [24]:
df.to_numpy()

array([[ 1,  5,  2],
       [ 2,  4,  3],
       [ 3,  3,  5],
       [ 4,  2,  7],
       [ 5,  1, 11]])

# Ejercicio 54

Seleccionar columnas de un objeto `DataFrame` usando notación de slicing.

## Solución:

In [25]:
import pandas as pd
import numpy as np

In [26]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [27]:
df[['nombre', 'califico']]

nombre califico
a    Oliva       Sí
b  Daniela       No
c     Juan       Sí
d   Germán       No
e   Edward       No
f     Alex       Sí
g    Julio       Sí
h    Edgar       No
i    Angie       No
j   Irlesa       Sí

# Ejercicio 55

Obtener los ejes (atributos de filas y columnas) de un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [30]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [31]:
df.axes

[RangeIndex(start=0, stop=5, step=1), Index(['X', 'Y', 'Z'], dtype='object')]

In [33]:
indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [34]:
df.axes

[Index(['a', 'b', 'c', 'd', 'e'], dtype='object'),
 Index(['X', 'Y', 'Z'], dtype='object')]

# Ejercicio 56

Seleccionar filas y columnas específicas de un `DataFrame` con `iloc`.

## Solución:

In [35]:
import pandas as pd
import numpy as np

In [36]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [38]:
df.iloc[[1, 4], [0, 3]] # [0, 3] dentro del [1, 4]

nombre califico
b  Daniela       No
e   Edward       No

In [39]:
type(df.iloc[[1, 4], [0, 3]])

pandas.core.frame.DataFrame

# Ejercicio 57

Obtener la dimensión, el tamaño, y la forma de un objeto `DataFrame`.

## Solución:

In [40]:
import pandas as pd

In [41]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [42]:
df.ndim

2

In [43]:
df.size

15

In [44]:
df.shape

(5, 3)

# Ejercicio 58

Obtener la cantidad de memoria ocupada por cada columna en un `DataFrame`.

## Solución:

In [49]:
import pandas as pd
import numpy as np

In [50]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = [True, False, True, False, False, True, True, False, False, True]

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos  califico
a    Oliva     11.5         1      True
b  Daniela      8.0         3     False
c     Juan     15.5         2      True
d   Germán      NaN         3     False
e   Edward      8.0         2     False
f     Alex     19.0         3      True
g    Julio     13.5         1      True
h    Edgar      NaN         1     False
i    Angie      8.0         2     False
j   Irlesa     18.0         1      True

In [51]:
df.memory_usage()

Index       80
nombre      80
puntaje     80
intentos    80
califico    10
dtype: int64

# Ejercicio 59

Seleccionar filas y columnas a partir de sus nombres con `ix` y `loc`.

## Solución:

In [3]:
import pandas as pd
import numpy as np

In [4]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [6]:
df.ix[['a', 'i'], ['nombre', 'califico']]

AttributeError: 'DataFrame' object has no attribute 'ix'

In [57]:
df.loc[['a', 'i'], ['nombre', 'califico']]

nombre califico
a  Oliva       Sí
i  Angie       No

# Ejercicio 60

Comprobar si un objeto `DataFrame` está vacío con `empty`.

## Solución:

In [ ]:
import pandas as pd

In [65]:
df = pd.DataFrame({'X': []})

df

Empty DataFrame
Columns: [X]
Index: []

In [66]:
df.size

0

In [67]:
df.empty

True

In [68]:
import numpy as np

df = pd.DataFrame({'X': [np.nan]})

df

X
0 NaN

In [69]:
df.size

1

In [70]:
df.empty

False

# Ejercicio 61

Seleccionar los registros que tienen un campo nulo (nan).

## Solución:

In [71]:
import pandas as pd
import numpy as np

In [72]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [73]:
df[df['puntaje'].isnull()]

nombre  puntaje  intentos califico
d  Germán      NaN         3       No
h   Edgar      NaN         1       No

# Ejercicio 62

Crear un objeto `DataFrame` a partir de objetos `Series`.

## Solución:

In [ ]:
import pandas as pd

In [74]:
nombres = pd.Series(['Edward', 'Germán', 'Daniela'])
edades = pd.Series([27, 23, 19])

personas = {'nombre': nombres, 'edad': edades}

df = pd.DataFrame(data=personas)

df

nombre  edad
0   Edward    27
1   Germán    23
2  Daniela    19

# Ejercicio 63

Cambiar el tipo de dato de una o más columnas con `astype`.

## Solución:

In [ ]:
import pandas as pd

In [75]:
datos = {'X': [2, 3], 'Y': [5, 7]}

df = pd.DataFrame(data=datos)

df

X  Y
0  2  5
1  3  7

In [76]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   X       2 non-null      int64
 1   Y       2 non-null      int64
dtypes: int64(2)
memory usage: 164.0 bytes


In [77]:
df.dtypes

X    int64
Y    int64
dtype: object

In [78]:
df.astype({'X': 'int32'}).dtypes

X    int32
Y    int64
dtype: object

In [79]:
df.astype({'X': 'int32', 'Y': 'float64'}).dtypes

X      int32
Y    float64
dtype: object

# Ejercicio 64

Seleccionar registros donde los valores de una columna estén en un rango.

## Solución:

In [80]:
import pandas as pd
import numpy as np

In [81]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [82]:
df[df['puntaje'].between(13, 19)]

nombre  puntaje  intentos califico
c    Juan     15.5         2       Sí
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
j  Irlesa     18.0         1       Sí

# Ejercicio 65

Determinar los datos en un `DataFrame` que son nulos (NaN).

## Solución:

In [85]:
import pandas as pd
import numpy as np

In [86]:
nombres = ['Edward', 'Daniela', 'Germán']
edades = [5, 6, np.nan]
fecha_nacimiento = [pd.Timestamp('1990-01-03'), None, np.nan]

personas = {'nombre': nombres, 'edad': edades, 'fecha_nacimiento': fecha_nacimiento}

df = pd.DataFrame(data=personas)

df

nombre  edad fecha_nacimiento
0   Edward   5.0       1990-01-03
1  Daniela   6.0              NaT
2   Germán   NaN              NaT

In [87]:
df.isna()

nombre   edad  fecha_nacimiento
0   False  False             False
1   False  False              True
2   False   True              True

# Ejercicio 66

Encontrar los datos en un `DataFrame` que **no** son nulos (NaN).

## Solución:

In [88]:
import pandas as pd
import numpy as np

In [89]:
nombres = ['Edward', 'Daniela', 'Germán']
edades = [5, 6, np.nan]
fecha_nacimiento = [pd.Timestamp('1990-01-03'), None, np.nan]

personas = {'nombre': nombres, 'edad': edades, 'fecha_nacimiento': fecha_nacimiento}

df = pd.DataFrame(data=personas)

df

nombre  edad fecha_nacimiento
0   Edward   5.0       1990-01-03
1  Daniela   6.0              NaT
2   Germán   NaN              NaT

In [90]:
df.notna()

nombre   edad  fecha_nacimiento
0    True   True              True
1    True   True             False
2    True  False             False

# Ejercicio 67

Acceder un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [91]:
import pandas as pd

In [92]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [93]:
df.at['b', 'Y']

np.int64(4)

In [94]:
df.at['e', 'Z']

np.int64(11)

# Ejercicio 68

Modificar un valor o entrada específico de un objeto `DataFrame` usando `at`.

## Solución:

In [95]:
import pandas as pd

In [96]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [97]:
df.at['b', 'Y'] = 29

df

X   Y   Z
a  1   5   2
b  2  29   3
c  3   3   5
d  4   2   7
e  5   1  11

In [98]:
df.at['e', 'Z'] = 33

df

X   Y   Z
a  1   5   2
b  2  29   3
c  3   3   5
d  4   2   7
e  5   1  33

# Ejercicio 69

Modificar un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [99]:
import pandas as pd

In [101]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [102]:
df.loc['a', 'Z'] = 12

df

X  Y   Z
a  1  5  12
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

# Ejercicio 70

Acceder a un valor o entrada específico de un objeto `DataFrame` usando `loc`.

## Solución:

In [ ]:
import pandas as pd

In [103]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [104]:
df.loc['a', 'Z']

np.int64(2)

In [105]:
df.loc['d', 'Z']

np.int64(7)

# Ejercicio 71

Sumar los valores de una columna de un `DataFrame` con `sum`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [106]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [107]:
df['intentos'].sum()

np.int64(19)

# Ejercicio 72

Acceder a una entrada de un `DataFrame` utilizando índices numéricos con `iat`.

## Solución

In [108]:
import pandas as pd

In [109]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [110]:
df.iat[0, 2]

np.int64(2)

In [111]:
df.iat[4, 0]

np.int64(5)

In [112]:
df.iat[-1, 0]

np.int64(5)

In [113]:
df.iat[-1, -1]

np.int64(11)

# Ejercicio 73

Calcular el promedio de los valores de una columna de un `DataFrame`.

## Solución:

In [114]:
import pandas as pd
import numpy as np

In [115]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [119]:
df['puntaje'].mean()

np.float64(12.6875)

# Ejercicio 74

Insertar una nueva columna en una posición arbitraria con la función insert.

## Solución:

In [ ]:
import pandas as pd

In [121]:
datos = {'W': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

In [122]:
df.insert(1, 'X', 0)

df

W  X  Y   Z
a  1  0  5   2
b  2  0  4   3
c  3  0  3   5
d  4  0  2   7
e  5  0  1  11

In [123]:
df['X'] = [9, 5, 1, 7, 3]

df

W  X  Y   Z
a  1  9  5   2
b  2  5  4   3
c  3  1  3   5
d  4  7  2   7
e  5  3  1  11

# Ejercicio 75

 Agregar una nueva fila a un objeto `DataFrame` por medio de `loc`.

## Solución:

In [124]:
import pandas as pd
import numpy as np

In [125]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [126]:
df.loc['k'] = ['Sebastián', 17.9, 2, 'Sí']

df

nombre  puntaje  intentos califico
a      Oliva     11.5         1       Sí
b    Daniela      8.0         3       No
c       Juan     15.5         2       Sí
d     Germán      NaN         3       No
e     Edward      8.0         2       No
f       Alex     19.0         3       Sí
g      Julio     13.5         1       Sí
h      Edgar      NaN         1       No
i      Angie      8.0         2       No
j     Irlesa     18.0         1       Sí
k  Sebastián     17.9         2       Sí

# Ejercicio 76:

Recorrer cada una de las columnas de un objeto `DataFrame` con items.

## Solución:

In [ ]:
import pandas as pd

In [127]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [130]:
for etiqueta, contenido in df.items():
    print('Nombre columna:', etiqueta)
    print('Contenido:', contenido, sep='\n')

Nombre columna: X
Contenido:
a    1
b    2
c    3
d    4
e    5
Name: X, dtype: int64
Nombre columna: Y
Contenido:
a    5
b    4
c    3
d    2
e    1
Name: Y, dtype: int64
Nombre columna: Z
Contenido:
a     2
b     3
c     5
d     7
e    11
Name: Z, dtype: int64


# Ejercicio 77

Eliminar una fila de un `DataFrame` con la función `drop`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [131]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [134]:
df.drop('h')

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

# Ejercicio 78

Ordenar en modo descendente un `DataFrame` por diferentes columnas.

## Solución:

In [135]:
import pandas as pd
import numpy as np

In [136]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [137]:
df.sort_values(by=['nombre', 'puntaje'], ascending=[False, True])

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
g    Julio     13.5         1       Sí
c     Juan     15.5         2       Sí
j   Irlesa     18.0         1       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
h    Edgar      NaN         1       No
b  Daniela      8.0         3       No
i    Angie      8.0         2       No
f     Alex     19.0         3       Sí

In [140]:
df.sort_values(by=['nombre', 'puntaje'], ascending=[False, False])

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
g    Julio     13.5         1       Sí
c     Juan     15.5         2       Sí
j   Irlesa     18.0         1       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
h    Edgar      NaN         1       No
b  Daniela      8.0         3       No
i    Angie      8.0         2       No
f     Alex     19.0         3       Sí

# Ejercicio 79

Iterar con la función `itertuples` un `DataFrame` por filas a modo de tuplas nombradas.

## Solución:

In [141]:
import pandas as pd

In [142]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [143]:
for registro in df.itertuples():
    print(registro)

Pandas(Index='a', X=1, Y=5, Z=2)
Pandas(Index='b', X=2, Y=4, Z=3)
Pandas(Index='c', X=3, Y=3, Z=5)
Pandas(Index='d', X=4, Y=2, Z=7)
Pandas(Index='e', X=5, Y=1, Z=11)


In [144]:
for registro in df.itertuples():
    print(registro.X, registro.Z, registro.Y * registro.X)

1 2 5
2 3 8
3 5 9
4 7 8
5 11 5


# Ejercicio 80

Reemplazar valores de una columna con la función `map` de un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [147]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [148]:
df['califico'] = df['califico'].map({'Sí': True, 'No': False})

df

nombre  puntaje  intentos  califico
a    Oliva     11.5         1      True
b  Daniela      8.0         3     False
c     Juan     15.5         2      True
d   Germán      NaN         3     False
e   Edward      8.0         2     False
f     Alex     19.0         3      True
g    Julio     13.5         1      True
h    Edgar      NaN         1     False
i    Angie      8.0         2     False
j   Irlesa     18.0         1      True

# Ejercicio 81

Extraer y remover una columna por su nombre a través de la función `pop`.

## Solución:

In [ ]:
import pandas as pd

In [149]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [150]:
df.pop('Y')

a    5
b    4
c    3
d    2
e    1
Name: Y, dtype: int64

In [151]:
df

X   Z
a  1   2
b  2   3
c  3   5
d  4   7
e  5  11

# Ejercicio 82

Reemplazar valores de una columna por medio de la función `replace`.

## Solución:

In [ ]:
import pandas as pd
import numpy as np

In [157]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [158]:
df['nombre'] = df['nombre'].replace('Alex', 'Alexander')

df

nombre  puntaje  intentos califico
a      Oliva     11.5         1       Sí
b    Daniela      8.0         3       No
c       Juan     15.5         2       Sí
d     Germán      NaN         3       No
e     Edward      8.0         2       No
f  Alexander     19.0         3       Sí
g      Julio     13.5         1       Sí
h      Edgar      NaN         1       No
i      Angie      8.0         2       No
j     Irlesa     18.0         1       Sí

# Ejercicio 83

Mostrar los n primeros y últimos registros de un objeto `DataFrame` con `head` y `tail`.

## Solución:

In [ ]:
import pandas as pd

In [159]:
datos_1 = list(range(1, 101))
datos_2 = list(range(1, 101))

df = pd.DataFrame({'A': datos_1, 'B': datos_2})

df

A    B
0     1    1
1     2    2
2     3    3
3     4    4
4     5    5
..  ...  ...
95   96   96
96   97   97
97   98   98
98   99   99
99  100  100

[100 rows x 2 columns]

In [160]:
df.head()

A  B
0  1  1
1  2  2
2  3  3
3  4  4
4  5  5

In [161]:
df.head(10)

A   B
0   1   1
1   2   2
2   3   3
3   4   4
4   5   5
5   6   6
6   7   7
7   8   8
8   9   9
9  10  10

In [163]:
df.tail()

A    B
95   96   96
96   97   97
97   98   98
98   99   99
99  100  100

In [162]:
df.tail(10)

A    B
90   91   91
91   92   92
92   93   93
93   94   94
94   95   95
95   96   96
96   97   97
97   98   98
98   99   99
99  100  100

# Ejercicio 84

Validar con `isin` si un elemento se halla dentro de un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [164]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [165]:
df.isin([10, 3])

X      Y      Z
a  False  False  False
b  False  False   True
c   True   True  False
d  False  False  False
e  False  False  False

In [166]:
df.isin({'Z': [10, 3]})

X      Y      Z
a  False  False  False
b  False  False   True
c  False  False  False
d  False  False  False
e  False  False  False

# Ejercicio 85

Reemplazar valores que no cumplan una condición a través de la función `DataFrame.where`.

## Solución:

In [ ]:
import pandas as pd

In [167]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [168]:
df.where(df % 2 == 0, -df)

X  Y   Z
a -1 -5   2
b  2  4  -3
c -3 -3  -5
d  4  2  -7
e -5 -1 -11

# Ejercicio 86

Agregar una nueva fila a un `DataFrame` con notación de slicing.

## Solución:

In [169]:
import pandas as pd
import numpy as np

In [170]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [171]:
colores = ['Negro', 'Blanco', 'Azul', 'Verde', 'Rojo', 'Amarillo', 
           'Gris', 'Naranja', 'Lila', 'Café']

df['color'] = colores

df

nombre  puntaje  intentos califico     color
a    Oliva     11.5         1       Sí     Negro
b  Daniela      8.0         3       No    Blanco
c     Juan     15.5         2       Sí      Azul
d   Germán      NaN         3       No     Verde
e   Edward      8.0         2       No      Rojo
f     Alex     19.0         3       Sí  Amarillo
g    Julio     13.5         1       Sí      Gris
h    Edgar      NaN         1       No   Naranja
i    Angie      8.0         2       No      Lila
j   Irlesa     18.0         1       Sí      Café

# Ejercicio 87

Usar la función `mask` para modificar los valores que cumplan una condición.

## Solución:

In [ ]:
import pandas as pd

In [172]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [173]:
df.mask(df > 3)

X    Y    Z
a  1.0  NaN  2.0
b  2.0  NaN  3.0
c  3.0  3.0  NaN
d  NaN  2.0  NaN
e  NaN  1.0  NaN

In [174]:
df.mask(df > 3, 0)

X  Y  Z
a  1  0  2
b  2  0  3
c  3  3  0
d  0  2  0
e  0  1  0

# Ejercicio 88

 Iterar cada una de las filas de un `DataFrame` con la función `iterrows`.
 
 ## Solución:

In [9]:
import pandas as pd
import numpy as np

In [10]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [15]:
for indice, fila in df.iterrows():
    print(fila['nombre'], fila['califico'])

Oliva Sí
Daniela No
Juan Sí
Germán No
Edward No
Alex Sí
Julio Sí
Edgar No
Angie No
Irlesa Sí


# Ejercicio 89

Realizar consultas sobre los elementos de un objeto `DataFrame` con la función `query`.

## Solución:

In [ ]:
import pandas as pd

In [16]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [17]:
df.query('X > Y')

X  Y   Z
d  4  2   7
e  5  1  11

In [18]:
df[df.X > df.Y]

X  Y   Z
d  4  2   7
e  5  1  11

# Ejercicio 90

Realizar operaciones aritméticas básicas entre dos objetos `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [29]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [22]:
df.add(df) #suma su propio valor

X   Y   Z
a   2  10   4
b   4   8   6
c   6   6  10
d   8   4  14
e  10   2  22

In [21]:
df.sub(df)

X  Y  Z
a  0  0  0
b  0  0  0
c  0  0  0
d  0  0  0
e  0  0  0

In [23]:
df.mul(df)

X   Y    Z
a   1  25    4
b   4  16    9
c   9   9   25
d  16   4   49
e  25   1  121

In [24]:
df.div(df)

X    Y    Z
a  1.0  1.0  1.0
b  1.0  1.0  1.0
c  1.0  1.0  1.0
d  1.0  1.0  1.0
e  1.0  1.0  1.0

In [30]:
datos = {'X': [5, 4, 3, 2, 1], 'Y': [1, 2, 3, 4, 5], 'Z': [2, 3, 5, 7, 11]}

df2 = pd.DataFrame(datos, index=indices)

df2

X  Y   Z
a  5  1   2
b  4  2   3
c  3  3   5
d  2  4   7
e  1  5  11

In [26]:
df.add(df2)

X  Y   Z
a  6  6   4
b  6  6   6
c  6  6  10
d  6  6  14
e  6  6  22

In [27]:
df.sub(df2)

X  Y  Z
a -4  4  0
b -2  2  0
c  0  0  0
d  2 -2  0
e  4 -4  0

In [31]:
df.mul(df2)

X  Y    Z
a  5  5    4
b  8  8    9
c  9  9   25
d  8  8   49
e  5  5  121

In [32]:
df.div(df2)

X    Y    Z
a  0.2  5.0  1.0
b  0.5  2.0  1.0
c  1.0  1.0  1.0
d  2.0  0.5  1.0
e  5.0  0.2  1.0

# Ejercicio 91

Obtener como una lista Python los nombres de las columnas de un `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [33]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [36]:
type(df.columns.values) # returns the type of this object

numpy.ndarray

In [35]:
list(df.columns.values) # store multiple items in a single variable

['X', 'Y', 'Z']

# Ejercicio 92

Calcular el módulo sobre cada elemento de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [37]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [38]:
df.mod(2)

X  Y  Z
a  1  1  0
b  0  0  1
c  1  1  1
d  0  0  1
e  1  1  1

In [39]:
df.mod(5)

X  Y  Z
a  1  0  2
b  2  4  3
c  3  3  0
d  4  2  2
e  0  1  1

# Ejercicio 93

Renombrar las columnas de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [40]:
datos = {'Columna1': [1, 2, 3, 4, 5], 'Columna2': [5, 4, 3, 2, 1], 
         'Columna3': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

Columna1  Columna2  Columna3
a         1         5         2
b         2         4         3
c         3         3         5
d         4         2         7
e         5         1        11

In [41]:
renombrado_columnas = {'Columna1': 'X', 'Columna2': 'Y', 'Columna3': 'Z'}

df.rename(columns=renombrado_columnas)

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

# Ejercicio 94

Calcular el producto punto entre dos objetos `DataFrame` con la función `dot`.

## Solución:

In [ ]:
import pandas as pd

In [43]:
datos_1 = [[1, 2, 3, 4, 5], [5, 4, 3, 2, 1], [2, 3, 5, 7, 11]]

datos_2 = [[1, 2, 3], [9, 8, 7], [1, 1, 1], [3, 2, 1], [8, 5, 2]]

df_1 = pd.DataFrame(datos_1)
df_2= pd.DataFrame(datos_2)

In [44]:
df_1

0  1  2  3   4
0  1  2  3  4   5
1  5  4  3  2   1
2  2  3  5  7  11

In [45]:
df_2

0  1  2
0  1  2  3
1  9  8  7
2  1  1  1
3  3  2  1
4  8  5  2

In [50]:
df_1.dot(df_2) # multiplicacion matricial

0    1   2
0   74   54  34
1   58   54  50
2  143  102  61

In [49]:
df_2.dot(df_1)

0   1   2    3    4
0  17  19  24   29   40
1  63  71  86  101  130
2   8   9  11   13   17
3  15  17  20   23   28
4  37  42  49   56   67

# Ejercicio 95

Seleccionar registros donde se cumpla una condición para valores de una columna.

## Solución:

In [51]:
import pandas as pd
import numpy as np

In [56]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

print(jugadores)

df = pd.DataFrame(data=jugadores, index=indices)

df

{'nombre': ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 'Edgar', 'Angie', 'Irlesa'], 'puntaje': [11.5, 8, 15.5, nan, 8, 19, 13.5, nan, 8, 18], 'intentos': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1], 'califico': ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']}


nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [57]:
df.loc[df['califico'] == 'Sí']

nombre  puntaje  intentos califico
a   Oliva     11.5         1       Sí
c    Juan     15.5         2       Sí
f    Alex     19.0         3       Sí
g   Julio     13.5         1       Sí
j  Irlesa     18.0         1       Sí

# Ejercicio 96

Usar operadores relacionales para comparar elementos de objetos `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [58]:
datos_1 = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

datos_2 = {'X': [8, 9, 3, 2, 1], 'Y': [9, 2, 0, 1, 1], 'Z': [2, 4, 13, 2, 11]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [59]:
df_1

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [60]:
df_2

X  Y   Z
0  8  9   2
1  9  2   4
2  3  0  13
3  2  1   2
4  1  1  11

In [61]:
df_1 < df_2

X      Y      Z
0   True   True  False
1   True  False   True
2  False  False   True
3  False  False  False
4  False  False  False

In [62]:
df_1 > df_2

X      Y      Z
0  False  False  False
1  False   True  False
2  False   True  False
3   True   True   True
4   True  False  False

In [63]:
df_1 <= df_2

X      Y      Z
0   True   True   True
1   True  False   True
2   True  False   True
3  False  False  False
4  False   True   True

In [64]:
df_1 >= df_2

X      Y      Z
0  False  False   True
1  False   True  False
2   True   True  False
3   True   True   True
4   True   True   True

In [65]:
df_1 == df_2

X      Y      Z
0  False  False   True
1  False  False  False
2   True  False  False
3  False  False  False
4  False   True   True

# Ejercicio 97

Usar métodos de instancia de `DataFrame` para comparacionese de igualdad.

## Solución:

In [66]:
import pandas as pd

In [67]:
datos_1 = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

datos_2 = {'X': [8, 9, 3, 2, 1], 'Y': [9, 2, 0, 1, 1], 'Z': [2, 4, 13, 2, 11]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [68]:
df_1

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [69]:
df_2

X  Y   Z
0  8  9   2
1  9  2   4
2  3  0  13
3  2  1   2
4  1  1  11

In [70]:
df_1.eq(df_2)

X      Y      Z
0  False  False   True
1  False  False  False
2   True  False  False
3  False  False  False
4  False   True   True

In [71]:
df_1.ne(df_2)

X      Y      Z
0   True   True  False
1   True   True   True
2  False   True   True
3   True   True   True
4   True  False  False

# Ejercicio 98

Cambiar el orden de dos columnas de un objeto `DataFrame`.

## Solución:

In [ ]:
import pandas as pd

In [72]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

indices = ['a', 'b', 'c', 'd', 'e']

df = pd.DataFrame(data=datos, index=indices)

df

X  Y   Z
a  1  5   2
b  2  4   3
c  3  3   5
d  4  2   7
e  5  1  11

In [73]:
df = df[['X', 'Z', 'Y']]

df

X   Z  Y
a  1   2  5
b  2   3  4
c  3   5  3
d  4   7  2
e  5  11  1

# Ejercicio 99

Combinar dos `DataFrame` y dejar en cada columna el valor mínimo de la suma por columna.

## Solución:

In [ ]:
import pandas as pd

In [75]:
datos_1 = {'A': [1, 1], 'B': [5, 5]}
datos_2 = {'A': [3, 3], 'B': [2, 2]}

df_1 = pd.DataFrame(data=datos_1)
df_2 = pd.DataFrame(data=datos_2)

In [76]:
df_1

A  B
0  1  5
1  1  5

In [77]:
df_2

A  B
0  3  2
1  3  2

In [78]:
minimo_columna = lambda s1, s2: s1.sum() if s1.sum() < s2.sum() else s2.sum()

df_1.combine(df_2, minimo_columna)

A  B
0  2  4
1  2  4

# Ejercicio 100

Agregar una nueva fila a una `DataFrame` con el método `append`.

## Solución:

In [ ]:
import pandas as pd

In [88]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [89]:
nuevo_registro = {'X': 6, 'Y': 0, 'Z': 13}

df = df._append(nuevo_registro, ignore_index=True)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11
5  6  0  13

# Ejercicio 101

Aplicar una función a todos los elementos de un `DataFrame` con la función `apply`.

## Solución:

In [ ]:
import pandas as pd

In [96]:
datos = {'X': [1, 2, 3, 4, 5], 'Y': [5, 4, 3, 2, 1], 'Z': [2, 3, 5, 7, 11]}

df = pd.DataFrame(data=datos)

df

X  Y   Z
0  1  5   2
1  2  4   3
2  3  3   5
3  4  2   7
4  5  1  11

In [99]:
import numpy as np

df.apply(np.sqrt).round(3)

X      Y      Z
0  1.000  2.236  1.414
1  1.414  2.000  1.732
2  1.732  1.732  2.236
3  2.000  1.414  2.646
4  2.236  1.000  3.317

In [100]:
def cuadrado(x):
    return x * x

df.apply(cuadrado)

X   Y    Z
0   1  25    4
1   4  16    9
2   9   9   25
3  16   4   49
4  25   1  121

In [110]:
df.apply(np.sum, axis=0)

X    15
Y    15
Z    28
dtype: int64

In [107]:
df.apply(np.sum, axis=1)

0     8
1     9
2    11
3    13
4    17
dtype: int64

In [108]:
df.apply(lambda x: x * x)

X   Y    Z
0   1  25    4
1   4  16    9
2   9   9   25
3  16   4   49
4  25   1  121

# Ejercicio 102

Guardar los registros de un `DataFrame` en un archivo de formato **CSV**.

## Solución:

In [111]:
import pandas as pd
import numpy as np

In [112]:
nombre = ['Oliva', 'Daniela', 'Juan', 'Germán', 'Edward', 'Alex', 'Julio', 
          'Edgar', 'Angie', 'Irlesa']
puntaje = [11.5, 8, 15.5, np.nan, 8, 19, 13.5, np.nan, 8, 18]
intentos = [1, 3, 2, 3, 2, 3, 1, 1, 2, 1]
califico = ['Sí', 'No', 'Sí', 'No', 'No', 'Sí', 'Sí', 'No', 'No', 'Sí']

indices = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

jugadores = {'nombre': nombre, 'puntaje': puntaje, 'intentos': intentos, 
             'califico': califico}

df = pd.DataFrame(data=jugadores, index=indices)

df

nombre  puntaje  intentos califico
a    Oliva     11.5         1       Sí
b  Daniela      8.0         3       No
c     Juan     15.5         2       Sí
d   Germán      NaN         3       No
e   Edward      8.0         2       No
f     Alex     19.0         3       Sí
g    Julio     13.5         1       Sí
h    Edgar      NaN         1       No
i    Angie      8.0         2       No
j   Irlesa     18.0         1       Sí

In [119]:
df.to_csv('jugadores.csv', index=False)